### PR TITLE
feat(kg): harvest Etan's voice-session decisions (ctx-strip + question split)

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary
- Add `brainlayer.kg_session_harvest` plus `scripts/kg_session_harvest.py` to harvest raw voice-session decisions into dictionary answers and apply-safe KG decisions.
- Split Q1-Q6 dictionary answers by batch stem/order regardless of recorded action, preserving verbatim notes as answers.
- Strip `ctx-*` synthetic members from canonical suggestions, merge member lists, mixed maps, and notes; rederive a real canonical by chunk count when the recorded canonical is synthetic.
- Validate the clean output by importing `scripts/kg_cleanup_apply.py::load_decisions`; the harvester does not open or write the BrainLayer DB.

## Test plan
- [x] `python3 -m pytest tests/test_kg_session_harvest.py` — 7 passed
- [x] `python3 -m pytest tests/test_kg_session_harvest.py tests/test_kg_review_session.py tests/test_kg_cleanup.py` — 34 passed before review fix; harvester suite rerun after fix
- [x] `python3 -m ruff check src/brainlayer/kg_session_harvest.py scripts/kg_session_harvest.py tests/test_kg_session_harvest.py` — all checks passed
- [x] `python3 -m pytest` — 2585 passed, 48 skipped, 5 xfailed, 328 warnings in 265.12s
- [x] pre-push hook gate — pytest unit suite 2511 passed, 9 skipped, 77 deselected, 1 xfailed; MCP registration 3 passed; isolated eval/hook routing 40 passed; Bun stale-index test passed; FTS5 determinism shell regression passed

## Notes
- `python3 -m ruff check .` currently fails on 56 pre-existing lint issues in unrelated scripts; this branch's changed files are Ruff-clean.
- Pre-commit `coderabbit review --agent` hit the 180s timeout after emitting one minor finding about malformed batch-member errors. Fixed with explicit member validation plus a regression test.
- Review required; do not merge this PR yet.

---

> [!NOTE]
> **Medium Risk**
> Incorrect ctx stripping or canonical re-derivation could produce wrong merge/keep payloads for downstream KG apply, though the harvester never writes the DB and round-trips through the apply validator.
> 
> **Overview**
> Adds **`brainlayer.kg_session_harvest`** and **`scripts/kg_session_harvest.py`** to turn raw Etan voice-session batch + decisions JSON into dictionary Q1–Q6 answers (JSON/Markdown) and a **`kg-flag-decisions-v1`** file safe for **`kg_cleanup_apply`**.
> 
> Answers are taken from decision **notes** keyed by question stem/order, not from whether the row was merge/keep; those question clusters are **dropped** from the exported cleanup decisions. Contested-cluster decisions are scrubbed of **`ctx-*`** synthetic members (canonical, merge losers, `needs_v1_1` maps, and `MIXED:` note JSON); when the recorded canonical is synthetic, the real canonical is **re-picked by highest chunk count**. Output is checked by dynamically calling **`scripts/kg_cleanup_apply.load_decisions`**—no BrainLayer DB access.
> 
> **`tests/test_kg_session_harvest.py`** covers happy paths, ctx/note edge cases, validation failures, and CLI execution without `PYTHONPATH`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 07dd8e00ee6fe82507fc199ecba4774000a1f8e7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Add `kg_session_harvest` pipeline to strip ctx-members and extract question-split decisions
> - Adds [`src/brainlayer/kg_session_harvest.py`](https://github.com/EtanHey/brainlayer/pull/460/files#diff-8c999acb7618c5cc1a9ac15c8d9519c92d4e4bbb3d66c6916137b7f4d96662e5) with a `harvest_session` function that loads a session batch and raw decisions, extracts six dictionary questions, builds answers, strips `ctx-` members from merge/keep/needs_v1_1 items, sanitizes notes, and writes JSON and Markdown artifacts to disk.
> - Adds a [`scripts/kg_session_harvest.py`](https://github.com/EtanHey/brainlayer/pull/460/files#diff-eac6e189021172386d18c4f8295c8f7559d1d5e19bcd47bde5d748a4795ac0af) CLI wrapper that invokes `main` and exits with its status code.
> - Validates clean decisions by dynamically importing the sibling `scripts/kg_cleanup_apply.py` loader, reusing its schema check instead of duplicating it.
> - Adds a comprehensive test suite in [`tests/test_kg_session_harvest.py`](https://github.com/EtanHey/brainlayer/pull/460/files#diff-06b18c93d266fcdbdefb24a974e8f3d02542870d91c7447f418388452155ad4d) covering ctx stripping, note sanitization, per-category counts, error cases, and a CLI invocation test without `PYTHONPATH`.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 07dd8e0.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added session harvesting functionality with automatic data cleaning and validation.
  * Generates cleaned decision outputs and corresponding metadata exports.
  * Includes comprehensive input validation and error handling for malformed data.
  * CLI tool now available for session processing workflows.

* **Tests**
  * Added extensive test coverage for harvesting workflows, validation scenarios, and CLI execution.

### CODE REVIEW

File: src/brainlayer/kg_session_harvest.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+    return [by_number[number] for number in range(1, 7)]
+
+
+def _validate_decision_stems(decisions: dict[str, Any], clusters: list[dict[str, Any]]) -> None:
+    stems = {cluster["stem"] for cluster in clusters}
+    for field in ("merge", "keep", "skipped", "needs_v1_1"):
+        for item in decisions.get(field, []):
+            stem = item.get("stem")
+            if stem not in stems:
+                raise ValueError(f"unknown decision stem {stem!r} in {field}")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Handled in `1a6580c3`: `_load_decisions` now validates every item in `merge`, `keep`, `skipped`, and `needs_v1_1` is an object before `_validate_decision_stems` runs, so malformed input fails loudly with `ValueError` instead of `AttributeError`. Covered by `test_harvest_fails_loud_on_non_object_decision_item`.

File: src/brainlayer/kg_session_harvest.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+    spec.loader.exec_module(module)
+    return module.load_decisions(path)
+
+
+def _contains_ctx(value: Any) -> bool:
+    if isinstance(value, dict):
+        return any(_contains_ctx(key) or _contains_ctx(val) for key, val in value.items())
+    if isinstance(value, list):
+        return any(_contains_ctx(item) for item in value)
+    return isinstance(value, str) and "ctx-" in value
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in `1a6580c3`: `_contains_ctx` now only treats strings as leaked ctx values when `_is_ctx_id(value)` matches, instead of substring-matching any `ctx-` text. Added `test_harvest_allows_non_id_note_text_containing_ctx_token`, which failed before the fix and passes now.

[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in `1a6580c3`: the final leak guard now rejects only actual ctx-style IDs via `_is_ctx_id(value)`, so metadata/notes can mention ctx-stripping without being treated as synthetic members. Covered by `test_harvest_allows_non_id_note_text_containing_ctx_token`.

File: src/brainlayer/kg_session_harvest.py [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+
+def _clean_mixed_note(note: str) -> str:
+    prefix = "MIXED: "
+    if not note.startswith(prefix):
+        return note
+    payload, separator, suffix = note[len(prefix) :].partition("; ")
+    try:
+        member_map = json.loads(payload)
+    except json.JSONDecodeError:
+        return note
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in `e6df2d3f`: `MIXED:` notes now fail loudly unless the payload parses as a JSON object, so malformed structural notes cannot carry unstripped ctx member IDs through as plain note text. Covered by `test_harvest_fails_loud_on_malformed_mixed_note_with_ctx_member`.

File: src/brainlayer/kg_session_harvest.py [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+
+    original_loser_ids = [member.get("id") for member in item.get("members", []) if isinstance(member, dict)]
+    unknown_loser_ids = [
+        member_id for member_id in original_loser_ids if member_id not in members_by_id and not _is_ctx_id(member_id)
+    ]
+    if unknown_loser_ids:
+        raise ValueError(f"merge members {unknown_loser_ids!r} are not real members of {cluster['stem']!r}")
+    clean_member_ids = _dedupe([member_id for member_id in original_loser_ids if member_id in members_by_id])
+    if not clean_member_ids:
+        clean_member_ids = [member["id"] for member in real_members]
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in `e6df2d3f`: merge items must now include a members list with at least one real selected member after ctx-strip, and at least one real loser after excluding the canonical. The previous fallback to all real members was removed. Covered by `test_harvest_fails_loud_when_merge_selects_no_real_losers`.

File: src/brainlayer/kg_session_harvest.py [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+
+def _validate_members(stem: str, members: list[Any]) -> None:
+    for member in members:
+        if not isinstance(member, dict) or not all(isinstance(member.get(key), str) for key in ("id", "name", "type")):
+            raise ValueError(f"invalid member in cluster {stem!r}: {member!r}")
+
+
+def _load_decisions(path: str | Path) -> dict[str, Any]:
+    decisions = json.loads(Path(path).read_text(encoding="utf-8"))
+    if decisions.get("schema") != DECISIONS_SCHEMA:
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py:90]: Fixed in `e6df2d3f`: `_load_decisions` now validates the top-level JSON value is an object before reading `schema`, matching `_load_batch` fail-loud behavior. Covered by `test_harvest_fails_loud_on_non_object_decisions_file`.

File: src/brainlayer/kg_session_harvest.py [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+        raise ValueError("ctx- member leaked into clean decisions")
+    return clean
+
+
+def _clusters_by_stem(clusters: list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
+    by_stem: dict[str, dict[str, Any]] = {}
+    for cluster in clusters:
+        stem = cluster["stem"]
+        if stem in by_stem:
+            raise ValueError(f"duplicate cluster stem {stem!r}")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in `e6df2d3f`: harvester lookup maps now key clusters and decisions by `(category, stem)`, aligning with the review-session `category:stem` cluster contract while still preserving Q1-Q6 ordering. Covered by `test_harvest_keys_clusters_by_category_and_stem`.

File: src/brainlayer/kg_session_harvest.py [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+        "rows_merged_away": sum(len(item["members"]) for item in merge),
+        "keep": len(keep),
+        "explicit": sum(1 for item in exported if item.get("source") not in RULE_SOURCES),
+        "by_rule": sum(1 for item in exported if item.get("source") in RULE_SOURCES),
+    }
+
+
+def _per_category(merge: list[dict[str, Any]], keep: list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
+    per_category: dict[str, dict[str, Any]] = {}
+    for item in merge + keep:
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in `83e8a9d3`: per-category totals are now recomputed from the batch, not just emitted clean decisions, so skipped/undecided non-question clusters remain visible in summaries. Covered by `test_harvest_per_category_counts_skipped_and_undecided_non_question_clusters`.

File: src/brainlayer/kg_session_harvest.py [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+    spec.loader.exec_module(module)
+    return module.load_decisions(path)
+
+
+def _contains_ctx(value: Any) -> bool:
+    if isinstance(value, dict):
+        return any(_contains_ctx(key) or _contains_ctx(val) for key, val in value.items())
+    if isinstance(value, list):
+        return any(_contains_ctx(item) for item in value)
+    return _is_ctx_id(value)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py:526]: Fixed in `82367b15`: copied notes now strip only ctx IDs that are actual members of the current cluster, preserving unrelated `ctx-` wording such as `ctx-local` or top-level `ctx-strip` source metadata. Covered by `test_harvest_strips_embedded_ctx_member_ids_from_notes`.

File: src/brainlayer/kg_session_harvest.py [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+    members_by_id = {member["id"]: member for member in real_members}
+    if not real_members:
+        raise ValueError(f"cluster {cluster['stem']!r} has no real members after ctx-strip")
+
+    canonical = item.get("canonical")
+    canonical_id = canonical.get("id") if isinstance(canonical, dict) else None
+    if canonical_id in members_by_id:
+        clean_canonical = _member_ref(members_by_id[canonical_id])
+    elif canonical_id is None or _is_ctx_id(canonical_id):
+        clean_canonical = _member_ref(_canonical_by_chunks(real_members))
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in `41d87f73`: merge decisions now fail loud unless `canonical` is an object with a string `id`. Only explicit synthetic `ctx-*` canonical IDs are re-derived from real member chunk counts. Covered by `test_harvest_fails_loud_when_merge_canonical_is_missing` and `test_harvest_fails_loud_when_merge_canonical_is_malformed`; full pre-push gate passed before push.

File: src/brainlayer/kg_session_harvest.py [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+
+def _load_apply_decisions(path: Path) -> dict[str, Any]:
+    root = Path(__file__).resolve().parents[2]
+    validator_path = root / "scripts" / "kg_cleanup_apply.py"
+    spec = importlib.util.spec_from_file_location("brainlayer_kg_cleanup_apply_validator", validator_path)
+    if spec is None or spec.loader is None:
+        raise RuntimeError(f"could not import kg cleanup validator from {validator_path}")
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module.load_decisions(path)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py:518]: Handled in `2fbbc341`: `_load_apply_decisions` now documents that the harvester is repo-local tooling and intentionally imports the sibling `scripts/kg_cleanup_apply.py` validator instead of duplicating its schema. Kept the existing import path because the worker brief explicitly requires importing that validator.

File: tests/test_kg_session_harvest.py [hunk 11]
```diff
@@ -0,0 +1,510 @@
...
+            "--answers",
+            str(answers_path),
+            "--decisions",
+            str(clean_path),
+        ],
+        capture_output=True,
+        text=True,
+        env=env,
+        check=False,
+    )
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/test_kg_session_harvest.py]: Handled in `2fbbc341`: the CLI subprocess test now uses `timeout=30` and fails explicitly with `pytest.fail` on `subprocess.TimeoutExpired`. Verified with the targeted CLI test, focused KG suite, Ruff, and the full pre-push gate.

File: src/brainlayer/kg_session_harvest.py [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+        raise ValueError(f"question cluster {cluster['stem']!r} member name must be a string")
+    return name
+
+
+def _build_clean_decisions(
+    clusters: list[dict[str, Any]],
+    questions: list[dict[str, Any]],
+    decisions: dict[str, Any],
+) -> dict[str, Any]:
+    question_keys = {_cluster_key(cluster) for cluster in questions}
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/kg_session_harvest.py]: Fixed in 07dd8e00. Added a regression test for a non-empty dictionary-questions category plus a duplicate Q1 stem in another category, and changed clean-decision question-keying to exclude every Q1-Q6 stem from apply-safe output while answer extraction still uses the official six question rows. Local pre-push gate passed.
